panther_hardware_interfaces/src/panther_system.cpp

```diff
@@ -0,0 +1,516 @@
+// Copyright 2023 Husarion sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <panther_hardware_interfaces/panther_system.hpp>
+
+#include <rclcpp/logging.hpp>
+
+#include <hardware_interface/types/hardware_interface_type_values.hpp>
+
+#include <panther_hardware_interfaces/utils.hpp>
+
+// todo: add user variable to the script that will trigger DOUT4 instead of safety stop
+
+namespace panther_hardware_interfaces
+{
+
+using namespace std::literals;
+
+void PantherSystem::CheckJointSize()
+{
+  if (info_.joints.size() != kJointsSize) {
+    throw std::runtime_error(
+      "Wrong number of joints defined: "s + std::to_string(info_.joints.size()) + ", "s +
+      std::to_string(kJointsSize) + "expected."s);
+  }
+}
+
+void PantherSystem::SortJointNames()
+{
+  // Sort joints names - later hw_states and hw_commands are accessed by static indexes, so it
+  // is necessary to make sure that joints are in specific order and order of definitions in URDF
+  // doesn't matter
+  for (std::size_t i = 0; i < kJointsSize; i++) {
+    for (std::size_t j = 0; j < kJointsSize; j++) {
+      if (info_.joints[j].name.find(joint_order_[i]) != std::string::npos) {
+        joints_names_sorted_[i] = info_.joints[j].name;
+      }
+    }
+  }
+}
+
+void PantherSystem::CheckJointNames()
+{
+  for (std::size_t i = 0; i < kJointsSize; i++) {
+    if (joints_names_sorted_[i] == "") {
+      throw std::runtime_error(
+        joint_order_[i] +
+        " joint not defined (exactly one joint containing this string is required)"s);
+    }
+  }
+}
+
+void PantherSystem::SetInitialValues()
+{
+  // It isn't safe to set command to NaN - sometimes it could be interpreted as Inf (although it
+  // shouldn't). In case of velocity, I think that setting the initial value to 0.0 is the best
+  // option.
+  for (std::size_t i = 0; i < kJointsSize; i++) {
+    hw_commands_velocities_[i] = 0.0;
+    hw_states_positions_[i] = std::numeric_limits<double>::quiet_NaN();
+    hw_states_velocities_[i] = std::numeric_limits<double>::quiet_NaN();
+    hw_states_efforts_[i] = std::numeric_limits<double>::quiet_NaN();
+  }
```

Comment:
```suggestion
  hw_commands_velocities_.fill(std::numeric_limits<double>::quiet_NaN());
  hw_states_positions_.fill(std::numeric_limits<double>::quiet_NaN());
  hw_states_velocities_.fill(std::numeric_limits<double>::quiet_NaN());
  hw_states_efforts_.fill(std::numeric_limits<double>::quiet_NaN());
```
This can work in case the `std::array` will be accepted

Reply:
I like it, but still I prefer to use 0.0 to fill commands (IMO it is safer, as I described in the comment).